---
title: "The multiMiR user's guide"
author: Yuanbin Ru (ruyuanbin@gmail.com)
date: "`r doc_date()`"
package: "`r pkg_ver('BiocStyle')`"
output: 
  BiocStyle::html_document:
    toc: true
    code_folding: show
vignette: >
  %\VignetteIndexEntry{The multiMiR user's guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---



# Introduction

microRNAs (miRNAs) regulate expression by promoting degradation or repressing
translation of target transcripts. miRNA target sites have been catalogued in
databases based on experimental validation and computational prediction using a
variety of algorithms. Several online resources provide collections of multiple
databases but need to be imported into other software, such as R, for
processing, tabulation, graphing and computation. Currently available miRNA
target site packages in R are limited in the number of databases, types of
databases and flexibility. The R package *multiMiR*, with web server at
[http://multimir.ucdenver.edu](http://multimir.ucdenver.edu), is a
comprehensive collection of predicted and validated miRNA-target interactions
and their associations with diseases and drugs. *multiMiR* includes
several novel features not available in existing R packages: 1) compilation of
14 different databases, more than any other collection; 2) expansion of
databases to those based on disease annotation and drug response, in addition
to many experimental and computational databases; and 3) user-defined cutoffs
for predicted binding strength to provide the most confident selection. The
*multiMiR* package enables retrieval of miRNA-target interactions from
14 external databases in R without the need to visit all these databases.
Advanced users can also submit SQL queries to the web server to retrieve
results.


# Getting to know the multiMiR database

```{r annotate,echo=FALSE}
options(width=100)
```

```{r multimir_dbTables}
library(multiMiR)
db.tables = multimir_dbTables()
db.tables
```

The *multiMiR* web server (\url{http://multimir.ucdenver.edu}) hosts a
database containing miRNA-target interactions from external databases. The
package *multiMiR* provides functions to communicate with the
*multiMiR* web server and its database. Information from each external
database is stored in a table in the *multiMiR* database. To see a list
of the tables, we can use the `r multimir_dbTables` function.

To display the database schema, we can use the `r multimir_dbSchema`
function.

```{r dbSchema, eval=FALSE, echo=TRUE}
multimir_dbSchema()
```

```{sql, eval=FALSE}
--
-- Table structure for table `mirna`
--

DROP TABLE IF EXISTS `mirna`;
CREATE TABLE `mirna` (
  mature_mirna_uid INTEGER UNSIGNED AUTO_INCREMENT,  -- mature miRNA unique ID
  org VARCHAR(4) NOT NULL,                           -- organism abbreviation
  mature_mirna_acc VARCHAR(20) default NULL,         -- mature miRNA accession
  mature_mirna_id VARCHAR(20) default NULL,          -- mature miRNA ID/name
  PRIMARY KEY (mature_mirna_uid),
  KEY org (org),
  KEY mature_mirna_acc (mature_mirna_acc),
  KEY mature_mirna_id (mature_mirna_id)
);

--
-- Table structure for table `target`
--

DROP TABLE IF EXISTS `target`;
CREATE TABLE `target` (
  target_uid INTEGER UNSIGNED AUTO_INCREMENT,   -- target gene unique ID
  org VARCHAR(4) NOT NULL,                      -- organism abbreviation
  target_symbol VARCHAR(80) default NULL,       -- target gene symbol
  target_entrez VARCHAR(10) default NULL,       -- target gene Entrez gene ID
  target_ensembl VARCHAR(20) default NULL,      -- target gene Ensembl gene ID
  PRIMARY KEY (target_uid),
  KEY org (org),
  KEY target_symbol (target_symbol),
  KEY target_entrez (target_entrez),
  KEY target_ensembl (target_ensembl)
);

--
-- Table structure for table `mirecords`
--

DROP TABLE IF EXISTS `mirecords`;
CREATE TABLE `mirecords` (
  mature_mirna_uid INTEGER UNSIGNED NOT NULL,   -- mature miRNA unique ID
  target_uid INTEGER UNSIGNED NOT NULL,         -- target gene unique ID
  target_site_number INT(10) default NULL,      -- target site number
  target_site_position INT(10) default NULL,    -- target site position
  experiment VARCHAR(160) default NULL,         -- supporting experiment
  support_type VARCHAR(40) default NULL,        -- type of supporting experiment
  pubmed_id VARCHAR(10) default NULL,           -- PubMed ID
  FOREIGN KEY (mature_mirna_uid)
    REFERENCES mirna(mature_mirna_uid)
    ON UPDATE CASCADE ON DELETE RESTRICT,
  FOREIGN KEY (target_uid)
    REFERENCES target(target_uid)
    ON UPDATE CASCADE ON DELETE RESTRICT
);

......

(Please note that only three of the 19 tables are shown here for demonstration
purpose.)

```

The function `r multimir_dbInfo` will display information of external
miRNA and miRNA-target databases in *multiMiR*, including version,
release date, link to download the data, and the corresponding table in
*multiMiR*.

```{r multimir_dbInfo}
db.info = multimir_dbInfo()
db.info
```

Among the 14 external databases, eight (DIANA-microT-CDS, ElMMo, MicroCosm,
miRanda, miRDB, PicTar, PITA, and TargetScan) contain predicted miRNA-target
interactions. Three of them (miRecords, miRTarBase, and TarBase) have
experimentally validated miRNA-target interactions. The remaining three
databases (miR2Disease, Pharmaco-miR, and PhenomiR) contain miRNA-drug/disease
associations.

To see how many records are in these 14 external databases we refer to the
`r multimir_dbCount` function.

```{r multimir_dbCount}
db.count = multimir_dbCount()
db.count
apply(db.count[,-1], 2, sum)
```

The current version of *multiMiR* contains nearly 50 million records.

# List miRNAs, genes, drugs and diseases in the multiMiR database}

In addition to functions displaying database and table information, the
*multiMiR* package also provides the `r list.multimir` function
to list all the unique miRNAs, target genes, drugs, and diseases in the
*multiMiR* database.

```{r list.multimir}
miRNAs = list.multimir("mirna")
dim(miRNAs)
head(miRNAs)
genes = list.multimir("gene")
dim(genes)
head(genes)
drugs = list.multimir("drug")
dim(drugs)
head(drugs)
diseases = list.multimir("disease")
dim(diseases)
head(diseases)
```

The current version of *multiMiR* has 4210 miRNAs and 94792 target
genes from human, mouse, and rat, as well as 64 drugs and 223 disease terms.
Depending on the speed of your Internet connection, it may take a few minutes to
retrieve the large number of target genes.


# Use `r get.multimir` to query the multiMiR database

`r get.multimir` is the main function in the package to retrieve
predicted and validated miRNA-target interactions and their disease and drug
associations from the *multiMiR* database.

To get familiar with the parameters in `r get.multimir`, you can type
`r ?get.multimir` or `r help(get.multimir)` in R. In the next section,
many examples illustrate the use of the parameters.


# Examples of multiMiR queries

In this section a variety of examples are described on how to query the multiMiR
database.

## Example 1: Retrieve all validated target genes of a given miRNA

In the first example, we ask what genes are validated targets of hsa-miR-18a-3p.

```{r Example1}
# The default is to search validated interactions in human
example1 = get.multimir(mirna='hsa-miR-18a-3p', summary=TRUE)
names(example1)
# Detailed information of the validated miRNA-target interaction
head(example1$validated)
# Which interactions are supported by Luciferase assay?
example1$validated[grep("Luciferase", example1$validated[,"experiment"]),]
example1$summary[example1$summary[,"target_symbol"] == "KRAS",]
```

It turns out that *KRAS* is the only target validated by Luciferase
assay. The interaction was recorded in miRecords and miRTarBase and supported by
the same literature, whose PubMed ID is in column `r pubmed_id`. The summary (by
setting `r summary = TRUE` when calling `r get.multimir`) shows the
number of records in each of the external databases and the total number of
databases supporting the interaction.

## Example 2: Retrieve miRNA-target interactions associated with a given drug
## or disease

In this example we would like to know which miRNAs and their target genes are
associated with Cisplatin, a chemotherapy drug used in several cancers.

```{r Example2} 
example2 = get.multimir(disease.drug='cisplatin', table='disease.drug')
names(example2)
nrow(example2$disease.drug)
head(example2$disease.drug)
```

`r get.multimir` returns 53 miRNA-target pairs. For more information, we
can always refer to the published papers with PubMed IDs in column
`r paper_pubmedID`.

## Example 3: Select miRNAs predicted to target a gene

`r get.multimir` also takes target gene(s) as input. In this example we
retrieve miRNAs predicted to target *Gnb1* in mouse. For predicted
interactions, the default is to query the top 20\% predictions within each
external database, which is equivalent to setting parameters
`r predicted.cutoff = 20` and `r predicted.cutoff.type = 'p'` (for
percentage cutoff). Here we search the top 35% among all conserved and
nonconserved target sites.

```{r Example3_part1}
example3 <- get.multimir(org = "mmu", 
                         target  = "Gnb1",
                         table   = "predicted",
                         summary = TRUE,
                         predicted.cutoff      = 35,
                         predicted.cutoff.type = "p",
                         predicted.site        = "all")
names(example3)
head(example3$predicted)
head(example3$summary)
```

The records in `r example3$predicted` are ordered by scores from best to
worst within each external database. Once again, the summary option allows us to
examine the number of target sites predicted by each external database and the
total number of databases predicting the interaction.

Finally we examine how many predictions each of the databases has.

```{r Example3_part2}
apply(example3$summary[, 6:13], 2, function(x) sum(x > 0))
```

## Example 4: Select miRNA(s) predicted to target most, if not all, of
## the genes of interest

You may have a list of genes involved in a common biological process. It is
interesting to check whether some, or all, of these genes are targeted by the
same miRNA(s). Here we have four genes involved in chronic obstructive pulmonary
disease (COPD) in human and want to know what miRNAs target these genes by
searching the top 500,000 predictions in each external database.

```{r Example4_part1} 
example4 <- get.multimir(org='hsa',
                         target=c('AKT2', 'CERS6', 'S1PR3', 'SULF2'),
                         table='predicted',
                         summary=TRUE,
                         predicted.cutoff.type='n',
                         predicted.cutoff=500000)
```

Then we count the number of target genes for each miRNA.

```{r Example4_part2}
example4.counts <- addmargins(table(example4$summary[, 2:3]))
example4.counts <- example4.counts[-nrow(example4.counts), ]
example4.counts <- example4.counts[order(example4.counts[, 5], decreasing = TRUE), ]
head(example4.counts)
```

## Example 5: Retrieve interactions between a set of miRNAs and a set of genes

In this example, we profiled miRNA and mRNA expression in poorly metastatic
bladder cancer cell lines T24 and Luc, and their metastatic derivatives FL4 and
Lul2, respectively. We identified differentially expressed miRNAs and genes
between the metastatic and poorly metastatic cells. Let's load the data.

```{r Example5, eval=FALSE, echo=TRUE}
load(url("http://multimir.ucdenver.edu/bladder.rda"))
```

Variable `r DE.miRNA.up` contains 9 up-regulated miRNAs and variable
`r DE.entrez.dn` has 47 down-regulated genes in the two metastatic cell
lines. The hypothesis is that interactions between these miRNAs and genes whose
expression changed at opposite directions may play a role in cancer metastasis.
So we use `r multiMiR` to check whether any of the nine miRNAs could
target any of the 47 genes.

```{r Example5_part2, eval=TRUE, echo=TRUE}
# search all tables & top 10% predictions
example5 <- get.multimir(org = "hsa",
                         mirna = DE.miRNA.up,
                         target = DE.entrez.dn,
                         table = "all",
                         summary = TRUE,
                         predicted.cutoff.type = "p",
                         predicted.cutoff = 10)
```
```{r eval = FALSE}
# Searching diana_microt ...
# Searching elmmo ...
# Searching microcosm ...
# Searching miranda ...
# Searching mirdb ...
# Searching pictar ...
# Searching pita ...
# Searching targetscan ...
# Searching mirecords ...
# Searching mirtarbase ...
# Searching tarbase ...
# Searching mir2disease ...
# Searching pharmaco_mir ...
# Searching phenomir ...
```

In the result, there are 3 unique miRNA-gene pairs that have been validated.

```{r Example5_part3, eval=FALSE, echo=TRUE}
example5$validated
```
```{r eval=FALSE}
#     database mature_mirna_acc mature_mirna_id target_symbol target_entrez
# 1 mirtarbase     MIMAT0000087  hsa-miR-30a-5p          FDX1          2230
# 2 mirtarbase     MIMAT0000087  hsa-miR-30a-5p        LIMCH1         22998
# 3    tarbase     MIMAT0000087  hsa-miR-30a-5p          FDX1          2230
# 4    tarbase     MIMAT0000424     hsa-miR-128          NEK2          4751
# 5    tarbase     MIMAT0000087  hsa-miR-30a-5p        LIMCH1         22998
#    target_ensembl               experiment          support_type pubmed_id
# 1 ENSG00000137714               Proteomics Functional MTI (Weak)  18668040
# 2 ENSG00000064042 pSILAC//Proteomics;Other Functional MTI (Weak)  18668040
# 3 ENSG00000137714               Proteomics              positive          
# 4 ENSG00000117650               Microarray              positive          
# 5 ENSG00000064042               Proteomics              positive          
```

Two miRNAs are associated with bladder cancer in miR2Disease and PhenomiR.

```{r Example5_part4, eval=FALSE, echo=TRUE}
example5$disease.drug[grep("bladder", example5$disease.drug$disease_drug,
                           ignore.case=TRUE),]
```
```{r eval=FALSE}
#        database mature_mirna_acc mature_mirna_id target_symbol target_entrez
# 18  mir2disease     MIMAT0000418  hsa-miR-23b-3p            NA            NA
# 711    phenomir     MIMAT0000418  hsa-miR-23b-3p            NA            NA
# 311    phenomir     MIMAT0000449 hsa-miR-146a-5p            NA            NA
#     target_ensembl   disease_drug
# 18              NA bladder cancer
# 711             NA Bladder cancer
# 311             NA Bladder cancer
#                                               paper_pubmedID
# 18  2007. Micro-RNA profiling in kidney and bladder cancers.
# 711                                                 17826655
# 311                                                 19127597
```

The predicted databases predict 65 miRNA-gene pairs between the 9 miRNAs and 28
of the 47 genes.

```{r Example5_part5, eval=FALSE, echo=TRUE}
length(unique(example5$predicted$mature_mirna_id))
# [1] 9
```


```{r Example5_part6, eval=FALSE, echo=TRUE}
length(unique(example5$predicted$target_entrez))
@
```{r, eval=FALSE}
# [1] 28
```


```{r Example5_part7, eval=FALSE, echo=TRUE}
unique.pairs = unique(data.frame(miRNA.ID =
                                 as.character(example5$predicted$mature_mirna_id),
target.Entrez=as.character(example5$predicted$target_entrez)))
nrow(unique.pairs)
```
```{r, eval=FALSE}
# [1] 65
```


```{r Example5_part8, eval=FALSE, echo=TRUE}
head(unique.pairs)
```
```{r, eval = FALSE}
#         miRNA.ID target.Entrez
# 1 hsa-miR-182-5p          2017
# 2 hsa-miR-182-5p          1112
# 3 hsa-miR-30d-5p         22998
# 4 hsa-miR-30a-5p         22998
# 5 hsa-miR-30b-5p         22998
# 6 hsa-miR-182-5p          5962
```

Results from each of the predicted databases are already ordered by their scores
from best to worst.


```{r Example5_part9, eval=FALSE, echo=TRUE}
example5.split = split(example5$predicted, example5$predicted$database)
```




# Direct query to the database on the multiMiR web server

As shown previously, *get.multimir* is the main function to retrieve
information from the *multiMiR* database, which is hosted at
http://multimir.ucdenver.edu. The function builds one SQL query for every
external database that the user is going to search, submits the query to the web
server, and parses, combines, and summarizes results from the web server. For
advanced users, there are a couple ways to query the *multiMiR*
database without using the *multiMiR* package; but they have to be
familiar with SQL queries. In general, users are still advised to use the
`r get.multimir` function when querying multiple external databases in
*multiMiR*.

## Direct query on the web server

The *multiMiR* package communicates with the *multiMiR*
database via the script http://multimir.ucdenver.edu/cgi-bin/multimir.pl on the
web server. To show all tables in the database, the following URL can be typed
in the address bar of a web browser.


[](http://multimir.ucdenver.edu/cgi-bin/multimir.pl?query=show tables)


The content displayed on the resulting web page is similar with what is returned
by typing `r multimir\_dbTables()` in R. Once again, data from each of the
external databases is stored in a table in *multiMiR*. There are also tables for
miRNAs (table *mirna*) and target genes (table *target*).

To learn about the structure of a table (e.g. DIANA-microT data in table
*diana\_microt*), users can use URL


[](http://multimir.ucdenver.edu/cgi-bin/multimir.pl?query=describe diana\_microt)


Similar with Example 1, the following URL searches for validated target genes of
hsa-miR-18a-3p in miRecords.


[](http://multimir.ucdenver.edu/cgi-bin/multimir.pl?query=SELECT m.mature\_mirna\_acc, m.mature\_mirna\_id, t.target\_symbol, t.target\_entrez, t.target\_ensembl, i.experiment, i.support\_type, i.pubmed\_id FROM mirna AS m INNER JOIN mirecords AS i INNER JOIN target AS t ON (m.mature\_mirna\_uid=i.mature\_mirna\_uid and i.target\_uid=t.target\_uid) WHERE m.mature\_mirna\_id='hsa-miR-18a-3p')


As you can see, the query is long and searches just one of the three validated
tables in *multiMiR*. While in Example 1, one line of R command using the `r
get.multimir` function searches, combines and summarizes results from all three
validated external databases (miRecords, miRTarBase and TarBase).

## Direct query in R

The same direct queries we did above on the web server can be done in R as well.
To show all tables in the database:

```{r Direct_query1}
direct1 = search.multimir(query="show tables")
```

These are equivalent to using the `r multimir\_dbTables` function as showed
previously.

To show the structure of table *diana\_microt*:

```{r Direct_query2}
direct2 = search.multimir(query="describe diana_microt")
direct2
```

To search for validated target genes of hsa-miR-18a-3p in miRecords:

```{r Direct_query3}
direct3 <- search.multimir(query="select m.mature_mirna_acc, m.mature_mirna_id, t.target_symbol, t.target_entrez, t.target_ensembl, i.experiment, i.support_type, i.pubmed_id from mirna AS m INNER JOIN mirecords AS i INNER JOIN target AS t ON (m.mature_mirna_uid=i.mature_mirna_uid and i.target_uid=t.target_uid) WHERE m.mature_mirna_id='hsa-miR-18a-3p'")
direct3
```


# Session Info

```{r sessionInfo}
sessionInfo()
warnings()
```




